Convert old CLI standalone commands to laminas-cli commands

boesing opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: default X.Y.z branch or the oldest support X.Y.z
  * Bugfix: default X.Y.z branch or the oldest support X.Y.z
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: default X.Y.z branch or the oldest support X.Y.z
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | yes
| New Feature   | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding documentation?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

This converts the existing CLI commands to symfony commands which can be used with `laminas-cli`.
These should get automatically registered to upstream projects once `symfony/console` is available.

boesing committed:
feature: convert old CLI standalone commands to laminas-cli commands

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
feature: introduce laminas-component-installer compatibility

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

froschdesign commented:
@boesing 
I will take over the documentation. See #177

boesing committed:
qa: require `symfony/console` with at least v6 for phpunit

This avoids us to mock method return values due to the lack of native return types.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
bugfix: assert that realpath returns a `non-empty-string` for an existing file path

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented on src/Command/ConfigDumperCommand.php:
```diff
@@ -0,0 +1,169 @@
+use function is_writable;
+use function sprintf;
+
+final class ConfigDumperCommand extends Command
```
`symfony/console` is in `require-dev`, yet we have a direct dependency to it under `src/` here :thinking:

boesing replied:
Yeah, commands are not meant to be used like services.
I did the same in `laminas-cache` as I do not require upstream projects to have `laminas-cli`.

The `ConfigProvider` does not inject anything `laminas-cli` related until `symfony/console` is available (which would be the case when either a `laminas-cli` or `symfony/console` is used.

Therefore, I've added it as an optional dependency - but might worth to add it to `suggest` 🤔 

The services (which are meant to be re-used) can still be used without `symfony/console` tho.


So I guess we have two options:
1. move `symfony/console` to `require`
2. add `laminas-cli` to `suggest` and mark the commands as `@internal` which implies it should not be instantiated from outside

WDYT?

----

https://github.com/laminas/laminas-servicemanager/pull/176/files#diff-f0a2f25e61573ecaca3a489ef6fca6a6106b7f825ef5599ec374f5387db58f82R49-R54

boesing replied:
Since there was no further feedback, I'll go with option 2 here:

- [x] add laminas-cli to `suggest` and mark the commands as `@internal` which for CLI commands implies it should not be instantiated from outside

froschdesign replied:
Sorry for the late response, but I like the solution as in laminas-cache. 👍🏻

Xerkus replied:
For the record, I am fine with taken approach on following conditions:
- Commands are to be used in development only.
- Commands are not meant to be used in any kind of automation.

If it is something that would be used in CI/CD pipeline or in deployments it needs to be via satellite package with explicit dependencies.

boesing committed:
qa: suggest `laminas/laminas-cli` so that the CLI commands can be consumed

This also marks the existing CLI commands as internal as these are not meant to be instantiated from any project.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Xerkus commented on src/ConfigProvider.php:
```diff
@@ -0,0 +1,81 @@
+     */
+    private function getLaminasCliDependencies(): array
+    {
+        if (PHP_SAPI !== 'cli') {
+            return [];
+        }
```
This is problematic approach with config since config cache does not distinguish cli and running under any other SAPI.

Commands introduced here are for development where cache should not be enabled generally speaking, but this is the kind of approach that should not be used at all.

froschdesign replied:
@Xerkus 
> Commands introduced here are for development where cache should not be enabled generally speaking…

I would even go one step further: the commands will not work correctly when the cache is active.

> …but this is the kind of approach that should not be used at all.

What do you mean here exactly?

Xerkus replied:
> What do you mean here exactly?

I mean the config provider approach resulting in differing configuration based on some runtime condition. Like environment variable or SAPI type as in this instance. The condition does not have meaningful side effects here but it sets a precedent for the practice I can not recommend. Since it does not have meaningful effect I do not see a reason to use it in the first place.

I mean, application can choose to do it for a good reason but not generic component in its config provider.

boesing replied:
I can follow the reasoning here, esp. when it comes to config caches.
But again, we are talking about CLI-only tooling and it also verifies that another dependency has to be available.
Don't you think that this is okayish even tho it can have side-effects?

I mean, its not that we blindly merge code just because s1 says "but in PR XY some1 already did that as well and now I want my stuff in here as well based on runtime stuff".

We might have to carefully check what is being added but I cannot think of any CLI dependency which is provided by the servicemanager which needs to be available on production environments and thus, config cache should be disabled in those environments where the commands are in use anyways.

WDYT? Do you have better suggestions on how to provide these CLI commands in a more optional way?
i.e. via delegators?

Xerkus replied:
> I mean, its not that we blindly merge code just because s1 says "but in PR XY some1 already did that as well and now I want my stuff in here as well based on runtime stuff".

It is more of "they do it, that is how it is intended to be done" case. I would rather not have to check this at all and not give ideas to people doing third party stuff.

Because this approach in general triggers red flags in my mind I want a compelling reason to have it used rather than just "is okayish".

boesing replied:
As discussed via Slack:

- [x] I'll drop the `cli` SAPI check as everything would work without it as well.

boesing committed:
qa: remove SAPI check in `ConfigProvider` which is assumed as problematic

There can be too many side-effects, i.e. by having cached configs, etc. and thus, we decided to drop this.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>
